docs/source/scheduling-policies.rst

```diff
@@ -12,28 +12,95 @@ more information on how this these policies are enacted efficiently see
 Choosing Workers
 ----------------
 
-When a task transitions from waiting to a processing state we decide a suitable
-worker for that task.  If the task has significant data dependencies or if the
-workers are under heavy load then this choice of worker can strongly impact
-global performance.  Currently workers for tasks are determined as follows:
+When a task transitions from waiting to a processing state, we decide a suitable
+worker for that task. If the task has significant data dependencies or if the
+workers are under heavy load, then this choice of worker can strongly impact
+global performance. Similarly, the placement of root tasks affects performance
+of downstream computations, since it can determine how much data will need to be
+transferred between workers in the future. Different heuristics are used for these
+different scenarios:
 
-1.  If the task has no major dependencies and no restrictions then we find the
-    least occupied worker.
+Initial Task Placement
+~~~~~~~~~~~~~~~~~~~~~~
+
+We want neighboring root tasks to run on the same worker, since there's a
+good chance those neighbors will be combined in a downstream operation::
+
+      i       j
+     / \     / \
+    e   f   g   h
+    |   |   |   |
+    a   b   c   d
+    \   \  /   /
+         X
+
+In the above case, we want ``a`` and ``b`` to run on the same worker,
+and ``c`` and ``d`` to run on the same worker, reducing future
+data transfer. We can also ignore the location of ``X``, because assuming
+we split the ``a b c d`` group across all workers to maximize parallelism,
+then ``X`` will eventually get transferred everywhere.
+(Note that wanting to co-locate ``a b`` and ``c d`` would still apply even if
+``X`` didn't exist.)
+
+Calculating these cousin tasks directly by traversing the graph would be expensive.
+Instead, we use the task's TaskGroup, which is the collection of all tasks with the
+same key prefix. (``(random-a1b2c3, 0)``, ``(random-a1b2c3, 1)``, ``(random-a1b2c3, 2)``
+would all belong to the TaskGroup ``random-a1b2c3``.)
```

Comment:
I wish we could describe the heuristic without explaining TaskGroups (or at least that they were documented elsewhere), but I couldn't figure out how without something verbose and awkward.

@mrocklin in general I recognize that this addition is probably too much detail on the implementation, and could be much shorter and just describe the scheduler's objective of co-assigning related tasks. I'm not sure if we want to go into the details of the implementation here or not.